networkx/linalg/algebraicconnectivity.py

```diff
@@ -378,6 +374,9 @@ def algebraic_connectivity(
     --------
     laplacian_matrix
     """
+    global np
```

Comment:
This is some pretty sneaky code. I had to do a double take to figure out why you were doing this. I think I'd prefer if the `import numpy as np` line was just in every function rather than hoping that these functions are always called first. Its not the end of the world, but I place a sizable amount of weight on avoiding global variables whenever possible.

Reply:
I don't have a strong opinion on this. If we can avoid `np` global altogether, great. But if that's too much effort, this way is not terrible.

I wonder about the cost of this pattern though. With what we had, `import scipy` (and also `numpy`) was tried only once (per module). If you import it inside a function, then it is imported multiple times. If found, then the extra imports are nearly free, just dict lookup. But if _not_ found, then the filesystem will be checked at the start of each function, which could be onerous.

Reply:
The cost of a failed import will not be a limiting factor for speed with any of these functions. If speed is an issue they will already have installed numpy or scipy. The goal here is to make it fast to ignore these functions.  If you never call these functions, this will speed it up.

The problem with making a global np is only if someone is using np as a variable name in the `__main__` namespace. At least that is my understanding. I agree that global namespace interactions are to be avoided, but so are imports within a function instead of at the module level. Such is the nature of these import-time optimizations.

Reply:
I removed global np in favor of more import numpy as np calls inside the functions.
This only affects algebraicconnectivity.py